ipld/retriever.go

```diff
@@ -172,27 +176,26 @@ func (rs *retrieverSession) request(ctx context.Context, q *quadrant) {
 	wg := &sync.WaitGroup{}
 	wg.Add(size)
 
-	log.Debugw("requesting quadrant", "x", q.x, "y", q.y, "size", size)
+	log.Debugw("requesting quadrant", "axis", q.source, "x", q.x, "y", q.y, "size", size)
 	for i, root := range q.roots {
 		go func(i int, root cid.Cid) {
 			defer wg.Done()
+			// get the root node
 			nd, err := rs.dag.Get(ctx, root)
 			if err != nil {
-				log.Errorw("getting root", "root", root.String(), "err", err)
 				return
 			}
-			// TODO(@Wondertan): GetLeaves should return everything it was able to request even on error,
-			// 	so that we fill as much data as possible
-			// get leaves of left or right subtree
-			nds, err := GetLeaves(ctx, rs.dag, nd.Links()[q.x].Cid, make([]format.Node, 0, size))
-			if err != nil {
-				log.Errorw("getting all the leaves", "root", root.String(), "err", err)
-				return
-			}
-			// fill leaves into the square
-			for j, nd := range nds {
-				rs.square[q.index(i, j)] = nd.RawData()[1+NamespaceSize:]
-			}
+			// and go get shares of left or the right side of the whole col/row axis
+			// the left or the right side of the tree represent some portion of the quadrant
+			// which we put into the rs.square share-by-share by calculating shares' indexes using q.index
+			GetShares(ctx, rs.dag, nd.Links()[q.x].Cid, size, func(j int, share Share) {
+				// the R lock here is *not* to protect rs.square from multiple concurrent shares writes
+				// but to avoid races between share writes and repairing attempts
+				// shares are written atomically in their own slice slot
+				rs.squareLk.RLock()
```

Comment:
Super counter-intuitive, but great (ab)use of RWMutex :imp:

Reply:
Yeah, it's smart :) and avoids contention. I hope comment explains it well enough as you get it. The reason why it's needed is below.